tests/mlmodel_langchain/test_chain.py

```diff
@@ -0,0 +1,2023 @@
+# Copyright 2010 New Relic, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import langchain
+import openai
+import pytest
+from langchain.chains.openai_functions import (
+    create_structured_output_chain,
+    create_structured_output_runnable,
+)
+from langchain.prompts import ChatPromptTemplate
+from langchain.schema import BaseOutputParser
+from testing_support.fixtures import (
+    reset_core_stats_engine,
+    validate_custom_event_count,
+    validate_transaction_error_event_count,
+)
+from testing_support.validators.validate_custom_events import validate_custom_events
+from testing_support.validators.validate_error_trace_attributes import (
+    validate_error_trace_attributes,
+)
+from testing_support.validators.validate_transaction_metrics import (
+    validate_transaction_metrics,
+)
+
+from newrelic.api.background_task import background_task
+from newrelic.api.transaction import add_custom_attribute
+from newrelic.common.object_names import callable_name
+
+disabled_custom_insights_settings = {"custom_insights_events.enabled": False}
+
+_test_openai_chat_completion_messages = (
+    {"role": "system", "content": "You are a scientist."},
+    {"role": "user", "content": "What is 212 degrees Fahrenheit converted to Celsius?"},
+)
+chat_completion_recorded_events_uuid_message_ids = [
```

Comment:
Given how many expected event dictionaries need to be included for chains, it may improve the readability of these tests to move all of these expected values into a separate file like we do for Bedrock. This doesn't need to block this PR from being merged, but maybe we could look into doing some test cleanup before GA.

Reply:
That sounds like a good idea!